compiler/resolution/functionResolution.cpp

```diff
@@ -1628,6 +1628,166 @@ computeGenericSubs(SymbolMap &subs,
+  if (se == parent->baseExpr) {
+    if(parent->numActuals() == 1) {
+      VarSymbol* idxVar = toVarSymbol(toSymExpr(parent->get(1))->var);
+      if (idxVar && idxVar->immediate &&
+          idxVar->immediate->const_kind == NUM_KIND_INT)
+        {
+          int idxNum = idxVar->immediate->int_value();
+          if (1 <= idxNum && idxNum <= numArgs)
+            return idxNum;
+          // report an "index out of bounds" error otherwise?
+        }
+    }
+  }
+  return 0;
+}
+
+//
+// 'parent' corresponds to tuple(i); replace it with 'replFormal'.
+//
+// If 'parent' is the source of a prim_move into a temp 'dest',
+// then eliminate 'dest' altogether, replacing its uses with 'replFormal'.
+//
+// Q: Why not just keep move(dest,replFormal), which is simpler?
+// A: because 'move' does not preserv replFormal's ref-ness and const-ness,
+// so we will get errors when using dest on the l.h.s. and miss errors
+// when dest or deref(dest) is modified and it shouldn't.
+// Note that we are here before the code is resolved, so we do not
+// necessarily know whether 'replFormal' is a const and/or a ref.
+// To keep 'dest' around, instead of move we probably need to add
+// the AST equivalent of Chapel's "ref dest = replFormal;".
+//
+static void replaceVarargAccess(CallExpr* parent, SymExpr* se,
+                                ArgSymbol* replFormal, SymbolMap& tempReps)
+{
+  if (CallExpr* move = toCallExpr(parent->parentExpr))
+    if (move->isPrimitive(PRIM_MOVE)) {
+      Symbol* dest = toSymExpr(move->get(1))->var;
+      if (dest->hasFlag(FLAG_TEMP)) {
+        // Remove the def of 'dest'.
+        dest->defPoint->remove();
+        move->remove();
+        // Uses of 'dest' will come later. We replace them with 'replFormal'
+        // in substituteVarargTupleRefs() using 'tempReps'.
+        tempReps.put(dest, replFormal);
+        return;
+      }
+    }
+  // Otherwise, just replace 'parent'.
+  parent->replace(new SymExpr(replFormal));
+}
+
+//
+// Is 'parent' the call 'se'.size?
+//
+static bool isVarargSizeExpr(SymExpr* se, CallExpr* parent) {
+  if (parent->isNamed("size"))
+    // Is 'parent' a method call?
+    if (parent->numActuals() == 2)
+      if (SymExpr* firstArg = toSymExpr(parent->get(1)))
+        if (firstArg->var->type == dtMethodToken) {
+          // If this fails, ensure that isVarargSizeExpr still works correctly.
+          INT_ASSERT(parent->get(2) == se);
+          return true;
+        }
+  return false;
+}
+
+//
+// Does 'ast' contain use(s) of the vararg tuple 'formal'
+// that require 'formal' as a whole tuple?
+//
+// Keep in sync with substituteVarargTupleRefs().
+//
+static bool needVarArgTupleAsWhole(BlockStmt* ast, int numArgs,
+                                   ArgSymbol* formal)
+{
+  std::vector<SymExpr*> symExprs;
+  collectSymExprs(ast, symExprs);
+
+  for_vector(SymExpr, se, symExprs) {
+    if (se->var == formal) {
+      if (CallExpr* parent = toCallExpr(se->parentExpr)) {
```

Comment:
Sorry.  I know what is does in general.  What purpose is it serving in this
function?  I did spot any allocations in that function.


On May 11, 2016, at 1:28 PM, Vassily Litvinov <notifications@github.com<mailto:notifications@github.com>> wrote:


In compiler/resolution/functionResolution.cpp<https://github.com/chapel-lang/chapel/pull/3838#discussion_r62920615>:

> +
> +//
> +// Does 'ast' contain use(s) of the vararg tuple 'formal'
> +// that require 'formal' as a whole tuple?
> +//
> +// Keep in sync with substituteVarargTupleRefs().
> +//
> +static bool needVarArgTupleAsWhole(BlockStmt* ast, int numArgs,
> +                                   ArgSymbol* formal)
> +{
> +  std::vector<SymExpr*> symExprs;
> +  collectSymExprs(ast, symExprs);
> +
> +  for_vector(SymExpr, se, symExprs) {
> +    if (se->var == formal) {
> +      if (CallExpr* parent = toCallExpr(se->parentExpr)) {


SET_LINENO sets the file name/line number to be assign to all new AST nodes created (e.g. new SymExpr(...)) through the end of the lexical scope where SET_LINENO occurs and in all dynamically-scoped called made during this time. These file name/line number can be overridden by another SET_LINENO for the duration of its lexical scope and its dynamically-scoped calls.

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub<https://github.com/chapel-lang/chapel/pull/3838/files/20b76cba1018c724ee2d739174a99cfcfbb22a17#r62920615>